### ISSUE TITLE
virtual datasets: support for sparse index based mappings

### ISSUE THREAD

**Issue by dnelson86** [h5py/h5py#1124]
Hello,

I've been playing with #1060 @takluyver (HDF5 1.10.4, h5py 11da4d6), it's really awesome! I've been waiting a long time for this for our use case, and it is really coming together nicely.

One aspect of my use case is that I would like to link in virtual datasets which are sparse (i.e. have values for only a select subset of entries of a dataset). In this case, the indices corresponding to each value are also saved. So I'd like to change the vds_simple example into (dummy example):

```python
    # Create source files (1.h5 to 4.h5)
    for n in range(1, 5):
        with h5py.File('{}.h5'.format(n), 'w') as f:
            d = f.create_dataset('data', (100,), 'i4')
            d[:] = np.arange(100) + n

    # Assemble virtual dataset
    layout = h5py.VirtualLayout(shape=(4, 100), dtype='i4')

    for n in range(1, 5):
        filename = "{}.h5".format(n)
        vsource = h5py.VirtualSource(filename, 'data', shape=(100,))
        inds = np.arange(100)[::-1]
        layout[n - 1,inds] = vsource

    # Add virtual dataset to output file
    with h5py.File("VDS.h5", 'w', libver='latest') as f:
        f.create_virtual_dataset('data', layout, fillvalue=-5)
        print("Virtual dataset:")
        print(f['data'][:, :10])
```

Is this within the VDS spec, and not yet implemented in h5py, or is this request outside the spec? Currently I get:

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.6/site-packages/h5py/_hl/selections.py in _handle_simple(shape, args)
    450             try:
--> 451                 x,y,z = _translate_int(int(arg), length)
    452                 s = True

TypeError: only size-1 arrays can be converted to Python scalars

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-19-5e8d1051e801> in <module>
----> 1 util.dataConvert.vdsTest()

/u/dnelson/python/util/dataConvert.py in vdsTest()
   1951         vsource = h5py.VirtualSource(filename, 'data', shape=(100,))
   1952         inds = np.arange(100)[::-1]
-> 1953         layout[n - 1,inds] = vsource
   1954
   1955     # Add virtual dataset to output file

/usr/local/lib/python3.6/site-packages/h5py/_hl/vds.py in __setitem__(self, key, source)
    124
    125     def __setitem__(self, key, source):
--> 126         sel = SimpleSelection(self.shape)[key]
    127         self.sources.append(VDSmap(sel.id,
    128                                    source.path,

/usr/local/lib/python3.6/site-packages/h5py/_hl/selections.py in __getitem__(self, args)
    259             return self
    260
--> 261         start, count, step, scalar = _handle_simple(self.shape,args)
    262
    263         self._id.select_hyperslab(start, count, step)

/usr/local/lib/python3.6/site-packages/h5py/_hl/selections.py in _handle_simple(shape, args)
    452                 s = True
    453             except TypeError:
--> 454                 raise TypeError('Illegal index "%s" (must be a slice or number)' % arg)
    455         start.append(x)
    456         count.append(y)

TypeError: Illegal index "[99 98 97 96 95 94 93 92 91 90 89 88 87 86 85 84 83 82 81 80 79 78 77 76
 75 74 73 72 71 70 69 68 67 66 65 64 63 62 61 60 59 58 57 56 55 54 53 52
 51 50 49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28
 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4
  3  2  1  0]" (must be a slice or number)
> /usr/local/lib/python3.6/site-packages/h5py/_hl/selections.py(454)_handle_simple()
    452                 s = True
    453             except TypeError:
--> 454                 raise TypeError('Illegal index "%s" (must be a slice or number)' % arg)
    455         start.append(x)
    456         count.append(y)
```

**Comment by takluyver** [comment]
I think that should be possible, but I haven't tried it. I wrote the VDS construction code using `SimpleSelection`, which only allows slices and single indexes. Have a look at the `h5py._hl.selections` module and try to use a `FancySelection` object for the selection you want.

Here's the code that selects part of the `VirtualSource`:

https://github.com/h5py/h5py/blob/11da4d6813b7830cb6cb04478e7d017b341d064a/h5py/_hl/vds.py#L96-L99

Try doing the same with a `FancySelection` object. If that works, we can probably use the `select()` function instead of creating a `SimpleSelection` directly.

**Comment by dnelson86** [comment]
Thanks for the hint, I've changed `SimpleSelection` to `FancySelection` in `__setitem__` (instead of `__getitem__`, since I want to use the indices as the destination in the virtual layout, rather than the origin in the virtual source).

My new minimal test is:

```python
    # Create source file
    with h5py.File('1.h5', 'w') as f:
        d = f.create_dataset('data', (10,), 'i4')
        d[:] = np.arange(10)*10

    # Make target indices
    inds = [0,10,20,30,40,50,60,70,80,90]

    # Assemble virtual dataset
    layout = h5py.VirtualLayout(shape=(100,), dtype='i4')

    vsource = h5py.VirtualSource('1.h5', 'data', shape=(10,))
    layout[inds] = vsource
```

but this gives

```
/u/dnelson/python/util/dataConvert.py in vdsTestSimple()
   2015
   2016     vsource = h5py.VirtualSource('1.h5', 'data', shape=(10,))
-> 2017     layout[inds] = vsource
   2018
   2019     import pdb; pdb.set_trace()

AttributeError: 'NoneType' object has no attribute 'id'
> /usr/local/lib/python3.6/site-packages/h5py/_hl/vds.py(127)__setitem__()
    125     def __setitem__(self, key, source):
    126         sel = FancySelection(self.shape)[key]
--> 127         self.sources.append(VDSmap(sel.id,
    128                                    source.path,
    129                                    source.name,
```

which I don't quite understand.

Note that if I change `layout[inds] = vsource` to `layout[0:10] = vsource` then with this modification I get:

```
TypeError: Advanced selection inappropriate
> /usr/local/lib/python3.6/site-packages/h5py/_hl/selections.py(366)__getitem__()
    364             raise TypeError("Only one indexing vector or array is currently allowed for advanced selection")
    365         if len(sequenceargs) == 0:
--> 366             raise TypeError("Advanced selection inappropriate")
    367
    368         vectorlength = len(list(sequenceargs.values())[0])
```

**Comment by takluyver** [comment]
`FancySelection.__getitem__` doesn't have a `return self` at the end. So when you slice it you're getting `None`.

Try changing this line:

```python
sel = FancySelection(self.shape)[key]
```

To something like this:

```python
sel = FancySelection(self.shape)
sel[key]
```

(Mutating the selection object from `__getitem__` is a bit awkward, but that's what it is)

**Comment by dnelson86** [comment]
Thanks, yes this solves it and everything seems to work correctly. My more realistic test case is

```python
    import numpy as np
    np.random.seed(424242)

    size_full = 10
    size_sparse = 5
    num_files = 3

    # Create source files (1.h5 to 3.h5)
    for n in range(1, num_files+1):
        with h5py.File('{}.h5'.format(n), 'w') as f:
            data = np.arange(size_sparse) + 10**n
            f['data'] = data

            inds = np.arange(size_full)
            np.random.shuffle(inds) # create dummy indices which cover the full dset size
            inds = np.sort(inds[0:size_sparse]) # must be ascending
            f['inds'] = inds

    # Assemble virtual dataset
    layout = h5py.VirtualLayout(shape=(num_files, size_full), dtype=data.dtype)

    for n in range(1, num_files+1):
        filename = "{}.h5".format(n)
        with h5py.File(filename,'r') as f:
            inds = f['inds'][()]
            vsource = h5py.VirtualSource(f['data'])
            layout[n - 1, inds] = vsource

    # Add virtual dataset to output file
    with h5py.File("VDS.h5", 'w', libver='latest') as f:
        f.create_virtual_dataset('data', layout, fillvalue=0)
        print("Virtual dataset:")
        print(f['data'][:,:])
```

which produces correctly:

```
Virtual dataset:
[[   0   10    0   11    0   12   13    0    0   14]
 [   0  100    0  101    0  102  103  104    0    0]
 [   0 1000    0    0 1001    0 1002    0 1003 1004]]
```

The "advanced selection inappropriate" issue remains, I guess this motivates the `select()` generalization you mentioned?

**Comment by takluyver** [comment]
> The "advanced selection inappropriate" issue remains, I guess this motivates the select() generalization you mentioned?

Yup, `select()` should automatically choose the appropriate type of selection based on what kind of indices are used.

**Comment by dnelson86** [comment]
Great, I see just to summarize, my changes are (in vds.py) pretty simple:

```python
from .selections import SimpleSelection, FancySelection, select

class VirtualSource(object):
...
    def __getitem__(self, key):
        tmp = copy(self)
        tmp.sel = select(self.shape, key, dsid=None)
        return tmp

class VirtualLayout(object):
...
    def __setitem__(self, key, source):
        sel = select(self.shape, key, source.sel.id)
        self.sources.append(VDSmap(sel.id,
                                   source.path,
                                   source.name,
                                   source.sel.id))
```

I would definitely support if something like this could make it into master at some point.

This will fail if the selection on the VirtualSource `isinstance(arg, h5r.RegionReference)`.

Thanks again for the pointers.

**Comment by takluyver** [comment]
Would you like to make a pull request with your changes?

It would need a couple of tests added, probably in [this class](https://github.com/h5py/h5py/blob/2fcc4e86e1ac354a4aaa508834bc0208703f8fd3/h5py/tests/hl/test_vds/test_highlevel_vds.py#L233), and maybe also [this file](https://github.com/h5py/h5py/blob/master/h5py/tests/hl/test_vds/test_virtual_source.py). But even if you don't have time to work on the tests, the pull request would be useful.

**Comment by takluyver** [comment]
Closed by #1126.
